File: sonet-client/ios/Sonet/NativeApp/ViewModels/MultiStepSignupViewModel.swift
import Foundation
import SwiftUI
import Combine

@MainActor
class MultiStepSignupViewModel: ObservableObject {
    
    // MARK: - Published Properties
    @Published var displayName = ""
    @Published var username = ""
    @Published var password = ""
    @Published var passwordConfirmation = ""
    @Published var birthday = Calendar.current.date(byAdding: .year, value: -18, to: Date()) ?? Date()
    @Published var email = ""
    @Published var phoneNumber = ""
    @Published var bio = ""
    @Published var avatarImage: UIImage?
    @Published var selectedInterests: Set<String> = []
    @Published var contactsAccess = false
    @Published var locationAccess = false
    
    // MARK: - Alert States
    @Published var showSuccessAlert = false
    @Published var showErrorAlert = false
    @Published var errorMessage: String?
    
    // MARK: - Private Properties
    private let grpcClient: SonetGRPCClient
    private var cancellables = Set<AnyCancellable>()
    
    // MARK: - Initialization
    init(grpcClient: SonetGRPCClient = SonetGRPCClient(configuration: .development)) {
        self.grpcClient = grpcClient
    }
    
    // MARK: - Public Methods
    func completeSignup() async {
        do {
            // Validate all required fields
            guard validateAllFields() else {
                await showError("Please fill in all required fields")
                return
            }
            
            // Create user account
            let userProfile = try await grpcClient.registerUser(
                username: username,
                email: email,
                password: password
            )
            
            // Update user profile with additional information
            try await updateUserProfile(userProfile: userProfile)
            
            // Handle permissions
            await handlePermissions()
            
            // Show success
            await showSuccess()
            
        } catch {
            await showError("Failed to create account: \(error.localizedDescription)")
        }
    }
    
    // MARK: - Private Methods
    private func validateAllFields() -> Bool {
        // Basic validation
        guard !displayName.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return false }
        guard !username.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return false }
        guard password.count >= 8 else { return false }
        guard password == passwordConfirmation else { return false }
        
        // Age validation
        let calendar = Calendar.current
        let ageComponents = calendar.dateComponents([.year], from: birthday, to: Date())
        guard (ageComponents.year ?? 0) >= 13 else { return false }
        
        // Email validation
        guard !email.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return false }
        guard isValidEmail(email) else { return false }
        
        // Interests validation
        guard selectedInterests.count >= 3 else { return false }
        
        return true
    }
    
    private func isValidEmail(_ email: String) -> Bool {
        let emailRegex = "[A-Z0-9a-z._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,64}"
        let emailPredicate = NSPredicate(format: "SELF MATCHES %@", emailRegex)
        return emailPredicate.evaluate(with: email)
    }
    
    private func updateUserProfile(userProfile: UserProfile) async throws {
        // Create update request with all collected information
        let request = UpdateUserProfileRequest.with {
            $0.userID = userProfile.userID
            $0.displayName = displayName
            $0.bio = bio
            $0.interests = Array(selectedInterests)
            
            // Add phone number if provided
            if !phoneNumber.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                $0.phoneNumber = phoneNumber
            }
        }
        
        // Update profile
        _ = try await grpcClient.updateUserProfile(request: request)
        
        // Upload avatar if selected
        if let avatarImage = avatarImage {
            try await uploadAvatar(image: avatarImage, userId: userProfile.userID)
        }
    }
    
    private func uploadAvatar(image: UIImage, userId: String) async throws {
        // Convert UIImage to Data
        guard let imageData = image.jpegData(compressionQuality: 0.8) else {
            throw NSError(domain: "AvatarUpload", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to convert image to data"])
        }
        
        // Create upload request
        let request = UploadMediaRequest.with {
            $0.userID = userId
            $0.mediaType = .image
            $0.mediaData = imageData
            $0.fileName = "avatar_\(userId).jpg"
        }
        
        // Upload avatar
        _ = try await grpcClient.uploadMedia(request: request)
    }
    
    private func handlePermissions() async {
        // Request contacts access if enabled
        if contactsAccess {
            await requestContactsAccess()
        }
        
        // Request location access if enabled
        if locationAccess {
            await requestLocationAccess()
        }
    }
    
    private func requestContactsAccess() async {
        // Request contacts permission
        let status = await requestContactsPermission()
        
        if status {
            // Sync contacts with server
            await syncContacts()
        }
    }
    
    private func requestLocationAccess() async {
        // Request location permission
        let status = await requestLocationPermission()
        
        if status {
            // Get current location and send to server
            await sendLocationToServer()
        }
    }
    
    private func requestContactsPermission() async -> Bool {
        // This would integrate with Contacts framework
        // For now, return true as placeholder
        return true
    }
    
    private func requestLocationPermission() async -> Bool {
        // This would integrate with CoreLocation framework
        // For now, return true as placeholder
        return true
    }
    
    private func syncContacts() async {
        // This would sync contacts with the server
        // Implementation depends on your contacts sync service
    }
    
    private func sendLocationToServer() async {
        // This would send location to the server
        // Implementation depends on your location service
    }
    
    private func showSuccess() async {
        showSuccessAlert = true
    }
    
    private func showError(_ message: String) async {
        errorMessage = message
        showErrorAlert = true
    }
}

// MARK: - Extensions for gRPC Integration
extension MultiStepSignupViewModel {
    
    // Helper method to create user profile from collected data
    func createUserProfileRequest() -> RegisterUserRequest {
        return RegisterUserRequest.with {
            $0.username = username
            $0.email = email
            $0.password = password
            $0.displayName = displayName
            $0.birthday = ISO8601DateFormatter().string(from: birthday)
            $0.phoneNumber = phoneNumber.isEmpty ? "" : phoneNumber
            $0.bio = bio
            $0.interests = Array(selectedInterests)
        }
    }
    
    // Helper method to validate username availability
    func validateUsername() async -> Bool {
        // This would check with the server if username is available
        // For now, return true as placeholder
        return true
    }
    
    // Helper method to check email availability
    func validateEmail() async -> Bool {
        // This would check with the server if email is available
        // For now, return true as placeholder
        return true
    }
}

// MARK: - Mock Data for Development
extension MultiStepSignupViewModel {
    
    // Method to populate with mock data for testing
    func populateWithMockData() {
        displayName = "John Doe"
        username = "johndoe"
        password = "Password123!"
        passwordConfirmation = "Password123!"
        birthday = Calendar.current.date(byAdding: .year, value: -25, to: Date()) ?? Date()
        email = "john.doe@example.com"
        phoneNumber = "+1234567890"
        bio = "Passionate about technology and innovation. Always learning and growing!"
        selectedInterests = ["Technology", "Science", "Business", "Education", "Health"]
        contactsAccess = true
        locationAccess = true
    }
    
    // Method to clear all data
    func clearAllData() {
        displayName = ""
        username = ""
        password = ""
        passwordConfirmation = ""
        birthday = Calendar.current.date(byAdding: .year, value: -18, to: Date()) ?? Date()
        email = ""
        phoneNumber = ""
        bio = ""
        avatarImage = nil
        selectedInterests.removeAll()
        contactsAccess = false
        locationAccess = false
    }
}

File: sonet-android/app/src/main/java/xyz/sonet/app/viewmodels/MultiStepSignupViewModel.kt
package xyz.sonet.app.viewmodels

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import xyz.sonet.app.grpc.SonetGRPCClient
import xyz.sonet.app.grpc.SonetConfiguration
import xyz.sonet.app.grpc.proto.*
import java.time.LocalDate
import java.time.format.DateTimeFormatter

class MultiStepSignupViewModel(application: Application) : AndroidViewModel(application) {
    
    // MARK: - StateFlow Properties
    private val _displayName = MutableStateFlow("")
    val displayName: StateFlow<String> = _displayName.asStateFlow()
    
    private val _username = MutableStateFlow("")
    val username: StateFlow<String> = _username.asStateFlow()
    
    private val _password = MutableStateFlow("")
    val password: StateFlow<String> = _password.asStateFlow()
    
    private val _passwordConfirmation = MutableStateFlow("")
    val passwordConfirmation: StateFlow<String> = _passwordConfirmation.asStateFlow()
    
    private val _birthday = MutableStateFlow(LocalDate.now().minusYears(18))
    val birthday: StateFlow<LocalDate> = _birthday.asStateFlow()
    
    private val _email = MutableStateFlow("")
    val email: StateFlow<String> = _email.asStateFlow()
    
    private val _phoneNumber = MutableStateFlow("")
    val phoneNumber: StateFlow<String> = _phoneNumber.asStateFlow()
    
    private val _bio = MutableStateFlow("")
    val bio: StateFlow<String> = _bio.asStateFlow()
    
    private val _avatarImage = MutableStateFlow<String?>(null)
    val avatarImage: StateFlow<String?> = _avatarImage.asStateFlow()
    
    private val _selectedInterests = MutableStateFlow<Set<String>>(emptySet())
    val selectedInterests: StateFlow<Set<String>> = _selectedInterests.asStateFlow()
    
    private val _contactsAccess = MutableStateFlow(false)
    val contactsAccess: StateFlow<Boolean> = _contactsAccess.asStateFlow()
    
    private val _locationAccess = MutableStateFlow(false)
    val locationAccess: StateFlow<Boolean> = _locationAccess.asStateFlow()
    
    // MARK: - UI State
    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()
    
    private val _isSignupComplete = MutableStateFlow(false)
    val isSignupComplete: StateFlow<Boolean> = _isSignupComplete.asStateFlow()
    
    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()
    
    // MARK: - Private Properties
    private val grpcClient = SonetGRPCClient(getApplication(), SonetConfiguration.development)
    
    // MARK: - Public Methods
    fun updateDisplayName(name: String) {
        _displayName.value = name
    }
    
    fun updateUsername(name: String) {
        _username.value = name
    }
    
    fun updatePassword(pwd: String) {
        _password.value = pwd
    }
    
    fun updatePasswordConfirmation(pwd: String) {
        _passwordConfirmation.value = pwd
    }
    
    fun updateBirthday(date: LocalDate) {
        _birthday.value = date
    }
    
    fun updateEmail(email: String) {
        _email.value = email
    }
    
    fun updatePhoneNumber(phone: String) {
        _phoneNumber.value = phone
    }
    
    fun updateBio(bio: String) {
        _bio.value = bio
    }
    
    fun updateAvatarImage(imagePath: String?) {
        _avatarImage.value = imagePath
    }
    
    fun updateSelectedInterests(interests: Set<String>) {
        _selectedInterests.value = interests
    }
    
    fun updateContactsAccess(access: Boolean) {
        _contactsAccess.value = access
    }
    
    fun updateLocationAccess(access: Boolean) {
        _locationAccess.value = access
    }
    
    fun completeSignup() {
        viewModelScope.launch {
            try {
                _isLoading.value = true
                _errorMessage.value = null
                
                // Validate all required fields
                if (!validateAllFields()) {
                    _errorMessage.value = "Please fill in all required fields"
                    return@launch
                }
                
                // Create user account
                val userProfile = grpcClient.registerUser(
                    username = username.value,
                    email = email.value,
                    password = password.value
                )
                
                // Update user profile with additional information
                updateUserProfile(userProfile)
                
                // Handle permissions
                handlePermissions()
                
                // Mark signup as complete
                _isSignupComplete.value = true
                
            } catch (error: Exception) {
                _errorMessage.value = "Failed to create account: ${error.localizedMessage}"
            } finally {
                _isLoading.value = false
            }
        }
    }
    
    // MARK: - Private Methods
    private fun validateAllFields(): Boolean {
        // Basic validation
        if (displayName.value.trim().isEmpty()) return false
        if (username.value.trim().isEmpty()) return false
        if (password.value.length < 8) return false
        if (password.value != passwordConfirmation.value) return false
        
        // Age validation
        val age = LocalDate.now().year - birthday.value.year
        if (age < 13) return false
        
        // Email validation
        if (email.value.trim().isEmpty()) return false
        if (!isValidEmail(email.value)) return false
        
        // Interests validation
        if (selectedInterests.value.size < 3) return false
        
        return true
    }
    
    private fun isValidEmail(email: String): Boolean {
        val emailRegex = "[A-Z0-9a-z._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,64}".toRegex()
        return emailRegex.matches(email)
    }
    
    private suspend fun updateUserProfile(userProfile: UserProfile) {
        // Create update request with all collected information
        val request = UpdateUserProfileRequest.newBuilder()
            .setUserId(userProfile.userId)
            .setDisplayName(displayName.value)
            .setBio(bio.value)
            .addAllInterests(selectedInterests.value.toList())
            .build()
        
        // Update profile
        grpcClient.updateUserProfile(request)
        
        // Upload avatar if selected
        avatarImage.value?.let { imagePath ->
            uploadAvatar(imagePath, userProfile.userId)
        }
    }
    
    private suspend fun uploadAvatar(imagePath: String, userId: String) {
        try {
            // Read image file and convert to bytes
            val imageBytes = getApplication<Application>().assets.open(imagePath).readBytes()
            
            // Create upload request
            val request = UploadMediaRequest.newBuilder()
                .setUserId(userId)
                .setMediaType(MediaType.IMAGE)
                .setMediaData(com.google.protobuf.ByteString.copyFrom(imageBytes))
                .setFileName("avatar_$userId.jpg")
                .build()
            
            // Upload avatar
            grpcClient.uploadMedia(request)
            
        } catch (error: Exception) {
            // Handle avatar upload error gracefully
            // For now, just log the error
            error.printStackTrace()
        }
    }
    
    private suspend fun handlePermissions() {
        // Request contacts access if enabled
        if (contactsAccess.value) {
            requestContactsAccess()
        }
        
        // Request location access if enabled
        if (locationAccess.value) {
            requestLocationAccess()
        }
    }
    
    private suspend fun requestContactsAccess() {
        // Request contacts permission
        val status = requestContactsPermission()
        
        if (status) {
            // Sync contacts with server
            syncContacts()
        }
    }
    
    private suspend fun requestLocationAccess() {
        // Request location permission
        val status = requestLocationPermission()
        
        if (status) {
            // Get current location and send to server
            sendLocationToServer()
        }
    }
    
    private suspend fun requestContactsPermission(): Boolean {
        // This would integrate with Android permissions system
        // For now, return true as placeholder
        return true
    }
    
    private suspend fun requestLocationPermission(): Boolean {
        // This would integrate with Android permissions system
        // For now, return true as placeholder
        return true
    }
    
    private suspend fun syncContacts() {
        // This would sync contacts with the server
        // Implementation depends on your contacts sync service
    }
    
    private suspend fun sendLocationToServer() {
        // This would send location to the server
        // Implementation depends on your location service
    }
    
    // MARK: - Helper Methods
    fun createUserProfileRequest(): RegisterUserRequest {
        return RegisterUserRequest.newBuilder()
            .setUsername(username.value)
            .setEmail(email.value)
            .setPassword(password.value)
            .setDisplayName(displayName.value)
            .setBirthday(birthday.value.format(DateTimeFormatter.ISO_LOCAL_DATE))
            .setPhoneNumber(phoneNumber.value.ifEmpty { "" })
            .setBio(bio.value)
            .addAllInterests(selectedInterests.value.toList())
            .build()
    }
    
    suspend fun validateUsername(): Boolean {
        // This would check with the server if username is available
        // For now, return true as placeholder
        return true
    }
    
    suspend fun validateEmail(): Boolean {
        // This would check with the server if email is available
        // For now, return true as placeholder
        return true
    }
    
    fun clearError() {
        _errorMessage.value = null
    }
    
    fun resetSignupState() {
        _isSignupComplete.value = false
        _errorMessage.value = null
        _isLoading.value = false
    }
}

// MARK: - Mock Data for Development
fun MultiStepSignupViewModel.populateWithMockData() {
    updateDisplayName("John Doe")
    updateUsername("johndoe")
    updatePassword("Password123!")
    updatePasswordConfirmation("Password123!")
    updateBirthday(LocalDate.now().minusYears(25))
    updateEmail("john.doe@example.com")
    updatePhoneNumber("+1234567890")
    updateBio("Passionate about technology and innovation. Always learning and growing!")
    updateSelectedInterests(setOf("Technology", "Science", "Business", "Education", "Health"))
    updateContactsAccess(true)
    updateLocationAccess(true)
}

fun MultiStepSignupViewModel.clearAllData() {
    updateDisplayName("")
    updateUsername("")
    updatePassword("")
    updatePasswordConfirmation("")
    updateBirthday(LocalDate.now().minusYears(18))
    updateEmail("")
    updatePhoneNumber("")
    updateBio("")
    updateAvatarImage(null)
    updateSelectedInterests(emptySet())
    updateContactsAccess(false)
    updateLocationAccess(false)
}